lib/spack/spack/cmd/providers.py

```diff
@@ -35,11 +36,47 @@
 
 
 def setup_parser(subparser):
+    subparser.epilog = 'If called without argument returns ' \
+                       'the list of all valid virtual packages'
     subparser.add_argument(
-        'vpkg_spec', metavar='VPACKAGE_SPEC', nargs=argparse.REMAINDER,
-        help='find packages that provide this virtual package')
+        'virtual_package',
+        nargs='*',
+        help='find packages that provide this virtual package'
+    )
 
 
 def providers(parser, args):
-    for spec in spack.cmd.parse_specs(args.vpkg_spec):
-        colify(sorted(spack.repo.providers_for(spec)), indent=4)
+    valid_virtuals = sorted(spack.repo.provider_index.providers.keys())
+    valid_virtuals_str = ''
+    if not sys.stdout.isatty():
+        buffer = six.StringIO()
+        colify.colify(valid_virtuals, output=buffer)
+        valid_virtuals_str = buffer.getvalue()
+    else:
+        valid_virtuals_str += 'Virtual packages:\n\t'
+        valid_virtuals_str += ', '.join(valid_virtuals)
+
+    # If called without arguments, list all the virtual packages
+    if not args.virtual_package:
+        print(valid_virtuals_str)
+        return
```

Comment:
I feel like this just keeps getting more complicated. Can't we replace all of this with:
```python
def providers(parser, args):
    valid_virtuals = sorted(spack.repo.provider_index.providers.keys())

    # If called without arguments, list all the virtual packages
    if not args.virtual_package:
        if sys.stdout.isatty():
            tty.msg('Virtual packages:')
        colify.colify(valid_virtuals)  
```
This is like a one-line change from what is already in `spack list`.

Reply:
Hmm, I tried that, but I don't like very much the output on a tty:
```
$ spack providers
Virtual packages:
D
awk
blas
daal
elf
gl
glu
golang
ipp
java
jpeg
lapack
mkl
mpe
mpi
opencl
openfoam
pil
pkgconfig
scalapack
szip
tbb
```

Reply:
The other cumbersome part is raising an exception with this list in the error message. `colify` accepts a stream, and that requires to use a `StringIO` to construct the message string.

Reply:
Hmm, that doesn't look very colified... I don't know as much about this section of Spack. I wonder why it works for `spack list` but not `spack providers`. @tgamblin?

Reply:
Ok, got it. Using a `StringIO` changes the default behavior of `colify`. I'll try to simplify this tomorrow.

Reply:
@adamjstewart Good enough to be merged and close #1355?